mmf/utils/inference.py

```diff
@@ -0,0 +1,96 @@
+from mmf.common.registry import registry  # noqa
+from mmf.common.sample import Sample, SampleList
+from mmf.utils.build import build_encoder, build_model, build_processors
+from mmf.utils.checkpoint import load_pretrained_model
+from mmf.utils.general import get_current_device
+from PIL import Image
+
+
+class Inference:
+    def __init__(self, checkpoint_path=None):
+        self.checkpoint = checkpoint_path
+        if checkpoint_path is None:
+            self.args = self.get_parser().parse_args()
+            self.checkpoint = self.args.checkpoint
+        assert self.checkpoint is not None
+        self.processor, self.feature_extractor, self.model = self._build_model()
+
+    def get_parser(self):
+        parser = argparse.ArgumentParser()
+        parser.add_argument(
+            "--checkpoint", default=None, type=str, help="Inference config file"
+        )
+        return parser
+
+    def _build_model(self):
+        model_items = load_pretrained_model(self.checkpoint)
+        self.config = omegaconf.OmegaConf.create(model_items["full_config"])
+        processor = build_processors(self.config.processors)
+        feature_extractor = build_encoder(self.config.features.image_feature_encodings)
+        ckpt = model_items["checkpoint"]
+        model = build_model(model_items["config"])
+        model.load_state_dict(ckpt)
+
+        return processor, feature_extractor, model
+
+    def forward(self, image_path, text, image_format="path"):
+        text_output = self.processor["text_processor"](text)
+        if image_format == "path":
+            img = np.array(Image.open(image_path))
+        elif image_format == "url":
+            img = np.array(Image.open(requests.get(image_path, stream=True).raw))
+        img = torch.as_tensor(img)
+        max_detect = self.config.features.image_feature_encodings.params.max_detections
+
+        if self.config.features.image_feature_encodings.type == "frcnn":
+
+            image_preprocessed, sizes, scales_yx = self.processor["image_processor"](
+                img
+            )
+
+            image_output = self.feature_extractor(
+                image_preprocessed,
+                sizes=sizes,
+                scales_yx=scales_yx,
+                padding=None,
+                max_detections=max_detect,
+                return_tensors="pt",
+            )
+
+            image_output = image_output[0]
+        else:
+            image_preprocessed = self.processor["image_processor"](img)
+            image_output = self.feature_extractor(image_preprocessed)
+
+        sample = Sample(text_output)
+        sample.image_feature_0 = image_output
+
+        sample_list = SampleList([sample])
+        sample_list = sample_list.to(get_current_device())
+        self.model = self.model.to(get_current_device())
+
+        output = self.model(sample_list)
+
+        answers = output["scores"].argmax(dim=1)
+
+        answer = self.processor["answer_processor"].idx2word(answers[0])
```

Comment:
Currently this is not generic and can handle only VQA Q/A type input. What will be a good way to make it generic for all type of tasks(or majority of tasks)? cc @apsdehal

Reply:
I think we can define something like output_processor which defaults to argmax prediction processor and then can be replaced accordingly.